rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+
+When `tsh login` would usually generate a new private key in memory, it will instead use [`piv.GenerateKey`](https://pkg.go.dev/github.com/go-piv/piv-go/piv#YubiKey.GenerateKey) to generate a new key pair directly on the yubikey's PIV Authentication slot (`9a`). The certificate signing process will continue as usual, using the yubikey's public key, and the certs will be stored in `~/.tsh/keys/proxy` as usual.
+
+Neither the yubikey's public or private key will be stored, as the private key can't be exported and we can later access the public key by checking the yubikey's [slot cert](https://pkg.go.dev/github.com/go-piv/piv-go/piv#YubiKey.Attest).
+
+##### Yubikey PIN/Touch policies
+
+When we generate a new private key, we can provide it with a custom Touch and PIN policy. These policies will be enforced whenever the private key is used for signing or decrypting, which may occur multiple times for each `tsh` command.
+
+The available PIN policies are:
+ - `never`: never prompt the user for PIN
+ - `always`: always prompt the user for PIN
+ - `once`: prompt the user for PIN just once per session, where a session is an open connection to the yubikey (needs verification)
+   - Due to the current keystore implementation reloading keys multiple times for every command, this may still lead to the user being prompted for PIN multiple times in each tsh call. This can potentially be reworked, but it will most likely not be fixed in the initial implementation.
+
+The available Touch policies are:
+ - `never`: never prompt the user for touch
+ - `always`: always prompt the user for touch
+ - `cached`: cache user's touch for 15 seconds before prompting them for it again
+   - This should result in each `tsh` command requiring just one touch
+   - We can try to control when the user is prompted for touch by doing a no-op signature at the start of a `tsh` command. However, if the cached touch runs out during the command, the user may be prompted later in the command execution than desired.
+
+The user can provide `--yubikey-pin-policy=<never|always|once>` and `--yubikey-touch-policy=<never|always|cached>` to specify their desired policies. They will both default to `never`.
+
+##### PIN Prompt
+
+Once the key is generated, we will need to prompt the user to enter a new PIN for it. If the PIN policy is `never`, then this prompt will be skipped.
+
+TBD: need to flush out how this will interact with webauthn/etc.
+
+##### Management Key
+
+TBD
+
+### yubikey login session
+
+Since `tsh` cannot detect a private key on disk like usual, we'll need to provide a way for `tsh` to reaccess its login session in another way.
+
+On yubikey login, `tsh` will add `yubikey: true` to the user's `~/.tsh/proxy.yaml` file so that `tsh` knows to use the user's yubikey for subsequent `tsh` commands. If the yubikey or its private key is not found or does not match the certificates on file, the user will be receive an error or be prompted to re-login like usual.
```

Comment:
This could be tricky, as the yubikey could easily be unplugged (or plugged somewhere else). Can you poll the yubikeys and figure out if they have a good PIV key?

Another alternative is to store a path/ID of the yubikey under ~/.tsh, so when you poll for keys you know if you found a good one. That doesn't help with the yubi being moved to another machine, for example.

A final alternative is to get Yubi IDs from the server, after authn. That should work regardless of local state.

As with everything resident-key-like, I'd guess that the user can always delete the PIV slot without us knowing, so there's always this error situation too.